### DISCUSSION TITLE
Reuse last command using !!

### DISCUSSION THREAD

**Discussion by laguill** [xonsh/xonsh#5728] [Q&A]
Hello,
I know there is [bashisms](https://github.com/xonsh/xontrib-bashisms)
But I would like to better understand how xonsh history works.

How can I set an allias to get the `!!` command

Let's say I ran `apt update`
Then `sudo !!`

Here is what I tried
```xonsh
sudo @(__xonsh__.history[-1].cmd)
sudo: apt update : command not found

```
How should I do ?

**Comment by anki-code** [comment]
What about
```xsh
import shlex
sudo @(shlex.split(__xonsh__.history[-1].cmd))
```

**Comment by laguill** [comment]
It working great
One issue now is how do you create an aliases for that ?
Or should it be a function

I tried to create in my xonshrc but the command is run when xonshrc is sourced
```
import shlex
aliases["!!"]=@(shlex.split(__xonsh__.history[-1].cmd))
```

**Reply by anki-code** [reply]
When you run any command it will be parsed by xonsh parser so `!!` command will be parsed and you can't use it for alias i.e.
```xsh
aliases['..'] = 'cd ..'  # It's working because parser is ok with `..`.
aliases['!!'] = 'echo !!'  # It's not working because parser is not ok with `!!`. May be it's feature for parser. PR is welcome!
```
To achieve what you want you need to overwrite command by using event [`on_transform_command`](https://xon.sh/events.html#on-transform-command-cmd-str-str) e.g.
```xsh
@events.on_transform_command
def _bangbang(cmd, **kw):
    if cmd.startswith('!!'):  
        # Be careful with the condition! The modified command will be passed 
        # to `on_transform_command` event again and again until the event 
        # returns the same command. Newbies can make a mistake here and
        # end up with unintended looping.
        return __xonsh__.history[-1].cmd
    else:
        return cmd
```
